uc-crux-llvm/src/UCCrux/LLVM/FullType/Type.hs

```diff
@@ -0,0 +1,276 @@
+{-
+Module           : UCCrux.LLVM.FullType.Type
+Description      : 'FullType' is like a 'CrucibleTypes.CrucibleType' and a 'MemType.MemType'
+Copyright        : (c) Galois, Inc 2021
+License          : BSD3
+Maintainer       : Langston Barrett <langston@galois.com>
+Stability        : provisional
+
+A 'FullType' is like a 'CrucibleTypes.CrucibleType', but contains type
+information through pointer references. Alternatively, it\'s like a
+'MemType.MemType' that can be linked to a 'CrucibleTypes.CrucibleType' by
+type-level information.
+
+Using this machinery, heads off several sources of partiality/errors:
+
+* By passing a 'FullType' instead of a 'MemType.MemType' and a
+  'CrucibleTypes.CrucibleType', it becomes impossible to pass
+  incompatible/out-of-sync inputs.
+* When building a @RegValue@, using 'FullType' can help prevent ill-typed
+  pointers, or instances of metadata or @void@ appearing in invalid places.
+* There are a few sources of partiality in the 'MemType.MemType' to
+  'CrucibleTypes.TypeRepr' translation that can be avoided, specifically
+  ill-sized integer values.
+* 'FullType' distinguishes between pointers and pointer-widths integers.
+* 'FullType' distinguishes between code pointers and data pointers.
+-}
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE ImplicitParams #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE PolyKinds #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TemplateHaskell #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE TypeOperators #-}
+{-# OPTIONS_GHC -Wno-inaccessible-code #-}
+-- These come from TH-generated code
+{-# OPTIONS_GHC -Wno-overlapping-patterns #-}
+
+module UCCrux.LLVM.FullType.Type
+  ( type FullType (..),
+    FullTypeRepr (..),
+    PartTypeRepr (..),
+    MapToCrucibleType,
+    ToCrucibleType,
+    MapToBaseType,
+    ToBaseType,
+    isPtrRepr,
+    IsPtrRepr (..),
+  )
+where
+
+{- ORMOLU_DISABLE -}
+import           GHC.TypeLits (Nat)
+import           Data.Kind (Type)
+import           Data.Functor.Const (Const(Const))
+import           Data.Type.Equality (TestEquality(testEquality), (:~:)(Refl))
+import           Unsafe.Coerce (unsafeCoerce)
+
+import qualified Data.Parameterized.Context as Ctx
+import           Data.Parameterized.Context (Ctx)
+import           Data.Parameterized.Classes (OrdF(compareF), OrderingF(LTF, GTF, EQF))
+import           Data.Parameterized.NatRepr (NatRepr, type (<=))
+import qualified Data.Parameterized.TH.GADT as U
+
+import qualified Text.LLVM.AST as L
+
+import qualified What4.InterpretedFloatingPoint as W4IFP
+
+import qualified Lang.Crucible.Types as CrucibleTypes hiding ((::>))
+
+import           Lang.Crucible.LLVM.Extension (ArchWidth, LLVMArch)
+import qualified Lang.Crucible.LLVM.MemType as MemType
+{- ORMOLU_ENABLE -}
+
+-- | Type level only
+data FullType (m :: Type) where
```

Comment:
Yeah, will document. It forces compatibility with a specific instance of `ModuleTypes`, because they're constructed jointly with `m` hidden in an existential. It represents the fact that two `FullType`s or a `FullType` and a `ModuleTypes` were made from the same LLVM module (and at the same time).

Reply:
Added (lots of) commentary in https://github.com/GaloisInc/crucible/pull/673/commits/6e2cfe7a85e32d8e04b83f5833597b8bcf0fa5e3, let me know if that helps.